Add public Resolver API

euroelessar opened the pull request:
gRPC core part of #11685

This PR consists of multiple steps:
1. Exposing `grpc_uri` interface for providing parsed target to the resolver factory.
2. Exposing `grpc_addresses` interface for addresses collection (unsure if we want re-use existing grpc_lb_addresses or create new one)
3. Expose interface for registering user-defined `grpc_resolver_factory`.

Resolver API is observer-based to support both pull and push-based resolvers.

TODO before merging:
1. Tests for `grpc_uri` interface and ensure its proper propagation.
2. Update `build.yaml` by adding reference to new resolver and regenerate all affected build files

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/16617)
<!-- Reviewable:end -->

jhserrano commented on include/grpc/grpc.h:
```diff
@@ -201,6 +201,110 @@ GRPCAPI void grpc_channel_watch_connectivity_state(
+                                                const char* target,
+                                                void* reserved);
+
+/** Sets the value of the address at index \a index of \a addresses.
```
s/index/target/ ?

markdroth commented on test/core/client_channel/resolvers/custom_resolver_test.cc:
```diff
@@ -0,0 +1,359 @@
+    const char* scheme, grpc_combiner* combiner) {
+  grpc_core::ResolverFactory* factory =
+      grpc_core::ResolverRegistry::LookupResolverFactory(scheme);
+  grpc_channel_args channel_args = {0, nullptr};
```
If there are no args, then there's no need for this; just set `args.args` to nullptr.

markdroth commented on test/core/client_channel/resolvers/custom_resolver_test.cc:
```diff
@@ -0,0 +1,359 @@
+  }
+};
+
+TEST(CustomResolverTest, Failure) {
```
Suggest calling this something like `FailToInstantiateResolver`.

markdroth commented on test/core/client_channel/resolvers/custom_resolver_test.cc:
```diff
@@ -0,0 +1,359 @@
+      build_custom_resolver("custom", combiner);
+  GPR_ASSERT(resolver.get() != nullptr);
+  GPR_ASSERT(fake_resolver != nullptr);
+  // Test 1: normal resolution.
```
It looks like most of this was copied from fake_resolver_test.cc.  However, it's not clear to me that these tests make sense for the custom resolver -- we can probably do something much simpler here, because there's less state than there is in the fake resolver.

I think all we really need here is a couple of basic checks:

1. When the resolver plugin returns results to the observer, those results are returned to the client channel via `CustomResolver::NextLocked()`.

2. When `CustomResolver::RequestReresolutionLocked()` is called, the plugin's `request_reresolution()` method is called.

markdroth commented on src/core/ext/filters/client_channel/resolver/custom/custom_resolver.cc:
```diff
@@ -0,0 +1,363 @@
+  return args->args->uri;
+}
+
+grpc_addresses* grpc_addresses_create_with_capacity(size_t capacity,
```
The `grpc_addresses_*` functions should live in lb_policy_factory.h, where the `grpc_lb_addresses` struct (and its methods) are defined.

markdroth commented on src/core/ext/filters/client_channel/resolver/custom/custom_resolver.cc:
```diff
@@ -0,0 +1,363 @@
+
+class CustomResolverFactory;
+
+class CustomResolver : public Resolver {
```
Currently, each individual polling-based resolver needs to contain code to handle (a) automatically retrying with exponential backoff upon failure and (b) using a minimum reresolution interval to avoid hammering the name service.  However, because this functionality is actually common to all polling-based resolvers, we plan to eventually move it out of the individual resolver implementations and into the client channel code.  When we do that, we will remove this code from the individual resolver implementations, but we won't be able to do that for the plugins, since we won't control them.  So for now, I think we need to add that functionality here in the `CustomResolver` class, so that plugins can be written assuming that this functionality will be handled for them.  Later, when we move this functionality into the client channel code, we can remove it from here.

You can probably copy most of the implementation of this from src/core/ext/filters/client_channel/resolver/dns/native/dns_resolver.cc.

euroelessar replied:
Is it acceptable to update `PluginResolver` implementation to call `request_reresolution` based on the following factors?
* Direct `RequestReresolution` calls from gRPC core
* Periodic timer
* Exponential backoff on resolution failures

Can you also clarify how `Resolver` API will be changed in the future, please? I'm a bit confused how backoff it going to interact with push-based resolvers.

euroelessar commented on include/grpc/grpc.h:
```diff
@@ -201,6 +201,110 @@ GRPCAPI void grpc_channel_watch_connectivity_state(
+/**
+ * Get the scheme of \a uri.
+ */
+GRPCAPI const char* grpc_uri_get_scheme(const grpc_uri* uri);
```
Makes sense, I assume there is no code for converting grpc_uri back to the string yet?

markdroth replied:
No, but it would be trivial to add a method for that to the internal `grpc_uri` API.

euroelessar replied:
Is it valid to extend `ResolverArgs` by the target field? `ResolverRegistry` has the original one and can pass it to the factory for free without reconstruction.

markdroth replied:
That sounds fine.

euroelessar commented on include/grpc/grpc.h:
```diff
@@ -201,6 +201,110 @@ GRPCAPI void grpc_channel_watch_connectivity_state(
+/** Add backend address with \a target to \a addresses.
+ * Channel will connect to \a target directly.
+ * \a target is an uri with the ipv4, ipv6, or unix scheme. */
+GRPCAPI bool grpc_addresses_add_backend_address(grpc_addresses* addresses,
```
Does gRPC promise only API compatibility or also ABI one? For example, is it valid to ask user to pass a struct with fields, and add new fields in the future?

markdroth replied:
I think it's probably fine to add new fields to a struct in the future.  But I'd like @nicolasnoble to weigh in on this.

euroelessar commented on include/grpc/grpc.h:
```diff
@@ -201,6 +201,110 @@ GRPCAPI void grpc_channel_watch_connectivity_state(
+/**
+ * Set new list of \a addresses to \a observer.
+ */
+GRPCAPI void grpc_resolver_observer_set_addresses(
```
Thanks for the input, I've updated the documentation with information about lifetime of the observer, hopefully it helps. (observer is owned by application code, and it's responsibility of the application to destroy it at any moment of time, e.g. after destruction of the resolver itself).

I agree that interaction with gil-based languages is tricky in the context of callbacks. Hopefully observer-based API should eliminate the need in majority of such interactions. The remaining ones (request reresolution, destruction) can be moved to the background thread to not block on gil.

It's theoretically also possible to make completion queue-based API, but it looks more complicated for understanding and introduces some bootstrapping issues (i.e. which completion queue to use, especially assuming that none can be created at the plugin initialization time).

euroelessar commented:
@markdroth Thanks for the comments, I've addressed most of the comments. Please have a look at API again, once it starts making sense to both of us - I'll prepare a gRFC for it.

euroelessar commented:
@markdroth I've rethought API a bit to make it possible to implement a resolver without acquiring GIL in callbacks from grpc-core itself. Unfortunately it means that resolver creation function has to be async now. Please have a look and let me know if it does match your expectations better.

AspirinSJL commented:
What's the status of this PR?

mhaidrygoog commented:
Thanks for the PR! Please take a look at the update Review comments and conflicts.

jhaals commented:
Any guesstimate on when this would be finalized and released?

euroelessar commented:
Hi, we've slightly de-prioritized this effort on our side, but I'll be able to continue (rebase & address comments) early in Q2.

markdroth commented:
FYI, we've been making some recent changes to the internal resolver API to move it closer to the form that we're going to eventually want to expose.  I've just merged #17987, which adds a ResultHandler (much like the observer being proposed in this PR) to the API.  And I'm currently working on #18357, which changes the API to return the address list and service config separately from channel args.

There will be more changes coming over the course of Q2.

lidizheng commented:
@euroelessar Is there any update for the public Resolver API work?

jhaals commented:
It would be fantastic to get another update on the timeline of this feature.

vaelen commented:
Any update on this?

lovemyliwu commented:
@markdroth any update?

markdroth commented:
@euroelessar, are you still planning to finish this?

If not, we had talked about @lidizheng or @gnossen possibly picking up this work.

Falco20019 commented:
@markdroth Any update?

markdroth commented:
I think it's going to be on @lidizheng's plate to drive this forward.

anthogez commented:
> ny guesstimate on when this would be finalized and released?

Hi @markdroth / @lidizheng is this still being considered to be moved forward? Thanks

markdroth commented:
I doubt this particular PR will move forward, since the author seems to have abandoned it.  But I think this feature is still desirable.

@lidizheng has left our team, but @gnossen may want to pick up this work.

gnossen commented:
I've added this to the Python team backlog, but unless we have a very strong story for public demand, it's unlikely to get planned as a high priority any time soon.

yashykt commented:
@gnossen Feel free to close this if this PR does not need to remain open
